kong/kong.lua

```diff
@@ -157,6 +161,34 @@ function Kong.ssl_certificate()
   end
 end
 
+function Kong.balancer()
+  local addr = ngx.ctx.balancer_address
+  addr.tries = addr.tries + 1
+  if addr.tries > 1 then  
+    -- only call balancer on retry, first one is done in `core.access.before` which runs
+    -- in the ACCESS context and hence has less limitations than this BALANCER context
+    -- where the retries are executed
+    
+    -- record failure data
+    addr.failures = addr.failures or {}
+    local state, code = ngx_balancer.get_last_failure()
+    addr.failures[addr.tries-1] = { name = state, code = code }
```

Comment:
Ok, just wondering if this was still WIP or final. 

> If, on the first try, you call set_more_tries(limit - 1), with for example limit 3. Then after the first try and maximum 2 retries, nginx will give up. See set_more_retries

Maybe without the `-1`? I know the theory but wondering what really happens when nginx "gives up". What's the response being returned to the client?